EIPS/eip-2537.md

```diff
@@ -0,0 +1,301 @@
+- For the next `48` bytes (encoding of `c1` coefficient):
+  - Three most significant bits **must be** zero.
+
+##### Point of infinity encoding:
+
+Also referred as "zero point". For BLS12 curves point with coordinates `(0, 0)` (formal zeroes in Fp or Fp2) is *not* on the curve, so encoding of such point `(0, 0)` is used as a convention to encode point of infinity.
+
+##### Boolean encoding for subgroup checks:
+
+For subgroup checks it's required to encode whether it's required to run a subgroup check for the G1/G2 point or not. For this we encode a boolean as a *single byte* `0x00` for `false` and `0x01` for `true`.
+
+##### Encoding of scalars for multiplication operation:
+
+Scalar for multiplication operation is encoded as `32` bytes by performing BigEndian encoding of the corresponding (unsigned) integer. Corresponding integer is **not** required to be less than or equal than main subgroup size.
+
+#### ABI for operations
+
+##### ABI for G1 addition
+
+G1 addition call expects `192` bytes as an input that is interpreted as byte concatenation of two G1 points (`96` bytes each). Output is an encoding of addition operation result.
+
+Error cases:
+- Either of points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G1 multiplication
+
+G1 multiplication call expects `128` bytes as an input that is interpreted as byte concatenation of encoding of G1 point (`96` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result.
+
+Error cases:
+- Point being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G1 multiexponentiation
+
+G1 multiplication call expects `128*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G1 point (`96` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result.
+
+Error cases:
+- Any of G1 points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 addition
+
+G2 addition call expects `384` bytes as an input that is interpreted as byte concatenation of two G2 points (`192` bytes each). Output is an encoding of addition operation result.
+
+Error cases:
+- Either of points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiplication
+
+G2 multiplication call expects `224` bytes as an input that is interpreted as byte concatenation of encoding of G2 point (`192` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result.
```

Comment:
What is the size of the output?

Reply:
This comment applies to all of them, really

Reply:
Added in the last update too